tests/components/matrix/test_login.py

```diff
@@ -0,0 +1,55 @@
+"""Test MatrixBot._login."""
+
+import pytest
+
+from homeassistant.components.matrix import MatrixBot
+from homeassistant.exceptions import ConfigEntryAuthFailed, HomeAssistantError
+
+from tests.components.matrix.conftest import TEST_MXID, TEST_TOKEN
+
+
+async def test_login(matrix_bot: MatrixBot, mock_client):
+    """Test various good and bad login paths.
+
+    "Good" test password is configured by default from config.
+    """
+
+    # Test good login using good password and missing access token.
+    await matrix_bot._login()
+    assert matrix_bot._client.logged_in
+
+    # Test good login using a good password and bad stored access_token.
+    matrix_bot._access_tokens = {TEST_MXID: "WrongToken"}
```

Comment:
Why are you not using caplog like in `tests/components/matrix/test_join_rooms.py`?

Reply:
As part of aforementioned parametrized tests, I've included assertions of the contents of `caplog.messages` as well